### DISCUSSION TITLE
React IconLayer right click event not fireing

### DISCUSSION THREAD

**Discussion by IncredibleMrTim** [visgl/deck.gl#6032] [Q&A] [answered]
I have been trying to figure out how to get right click to work on IconLayer map marker.  I have tried the suggested way

eg.
```
onClick: (info, event) => console.log(event.rightcllick)
```

but this does not seem to be working for me.  Whenever I right-click no event is fired or my action listener fires from the map canvas instead (I have already tested removing the map listener I get the same results, no event)

After some further searching, I have found this in the docs.

https://deck.gl/docs/developer-guide/interactivity#calling-the-picking-engine-directly

however, when I fire the onclick i am unable to find any markers on the map and the `pickInfo` comes back as null.

I have had to slightly adapt the way it has been done as this is a class-based component so I am unable to use hooks

```javascript
    ...

    deckRef = React.createRef()

	deckClick = e => {
		const pickInfo = this.deckRef.current.pickObject({
			x: e.clientX,
			y: e.clientY,
			radius: 1,
			layerIds: ['icon-layer-TEST'],
		})

		console.log(pickInfo)
		console.log(e, e.clientX, e.clientY)
		console.log(this.deckRef.current)
	}


    render() {
		const { center, google, initialMapCenter, language, zoom } = this.props

		return (
			<Container
				className={MAP_CONTAINER_CLASS_NAME}
				id="map"
				onContextMenu={this.deckClick}
			>
				<DeckGL ref={this.deckRef}>
					<GoogleMap

    ...
```

**Comment by Pessimistress** [comment, marked as answer]
Make sure you [enable picking](https://deck.gl/docs/developer-guide/interactivity#enabling-picking).

**Reply by IncredibleMrTim** [reply]
Hi, thanks for the response.  I have already tried this

```javascript
			layers.push(
				new IconLayer({
					id: `icon-layer-TEST`,
					data: markerData,
					pickable: true,
					getIcon: d => {
						return {
							url: IconTemplatePoiSurvey(10),
							width: 30,
							height: 30,
						}
					},
					sizeScale: 1,
					getPosition: d => d.coordinates,
					getSize: d => 75,
					onClick: (info, event) => {
						console.log('T3', event)
					},
				})
```

when I do it using a ref i get this output I the handler function

```javascript
       // handles onContextMenu
	deckClick = e => {
		const pickInfo = this.deckRef.current.pickObject({
			y: e.clientX,
			x: e.clientY,
			radius: 1,
			layerIds: ['icon-layer-TEST'],
		})

		console.log(`PickInfo: ${pickInfo}`)
		console.log(`event: ${e}, clientX: ${e.clientX}, clientY: ${e.clientY}`)
		console.log(`deckRef: ${this.deckRef.current}`)
	}
````
![image](https://user-images.githubusercontent.com/5684161/127332761-9b7cc3f4-0264-4fa7-b974-96c1c8f9282c.png)

**Reply by Pessimistress** [reply]
So left click works?

`onClick` should just work if you call `preventDefault` in the onContextMenu handler.

**Reply by IncredibleMrTim** [reply]
Yes it works just fine for the left click.  if i add prevent default to the onContextMenu on the wrapper it makes no difference.

I also tried to remove the deckClick function and just have `e => e.preventDefault()` in there, but then I get no event at all for right click?

**Reply by Pessimistress** [reply]
[example](https://codesandbox.io/s/deckgl-right-click-r3fep)

**Reply by IncredibleMrTim** [reply]
Hey.  Thanks for the response..  So after playing with your example as well as this one...

https://codesandbox.io/s/deckgl-gmaps-example-forked-14hw1?file=/src/index.js

I see the difference is that in your example you use `DeckGL` whereas we are using the `GoogleMapsOverlay`.  I can see that your example does indeed work, however, when applying the same code to the `GoogleMapsOverlay` it does not work.

I did notice it work once, but I'm not sure if that was just a fluke as I was unable to recreate it again

![image](https://user-images.githubusercontent.com/5684161/127460950-1ca10290-e04d-461e-a2d7-44823550672e.png)

**Reply by Pessimistress** [reply]
GoogleMapsOverlay does not handle any user input on its own. It listens to Google Maps' `click` event which I believe only fires on left click. This is a limitation in the Google Maps API.

Your code in the original post was not complete. Why are you calling the picking methods on deckRef if you are not using the DeckGL component to render any layers?

**Reply by IncredibleMrTim** [reply]
this was due to me playing about with different ideas.

Is it possible to get the pickInfo on right clicking the map?

**Reply by Pessimistress** [reply]
Yes, you need to call the [picking methods](https://deck.gl/docs/api-reference/google-maps/google-maps-overlay#methods) on your overlay instance.
